### PR TITLE
Allow local users in LDAP mode

### PR DESCRIPTION
**Pull request by bgeuken**
User index page (only visible for admins):
user_index_page

Dropdown menu (part of the index page):
dropdown_menu_users

User edit page (admin view):
user_edit_page

### CODE REVIEW

**Comment by bgeuken** [comment]
[P4: (5) Allow local users in LDAP mode](https://trello.com/c/tj5UVn89/694-p4-5-allow-local-users-in-ldap-mode)

**Comment by bgeuken** [comment]
Let me know if you have a better name for the label of that attribute. I don't like 'Ignore external authorisation services' to much but can't think of a better one.

**Comment by Conan-Kudo** [comment]
@bgeuken Keep it simple, just say "Make this a local user".

**Comment by adrianschroeter** [comment]
Is it correct that we can not merge this into 2.9 branch after 2.9 release because the code would
crash without the migration?

So we would either need to change the code to accept a not migrated DB or we need to get it in before?

**Comment by bgeuken** [comment]
@adrianschroeter The database needs to be migrated to make this feature work.

Why wouldn't we be able to merge this into 2.9?

**Comment by bgeuken** [comment]
@Conan-Kudo I guess that would work assuming that we only show this attribute when LDAP mode is enabled :+1:

**Comment by adrianschroeter** [comment]
On Dienstag, 13. März 2018, 10:42:17 CET wrote Björn Geuken:
> @adrianschroeter The database needs to be migrated to make this feature work.
> 
> Why wouldn't we be able to merge this into 2.9?

because package updates must not break running instances. So we can not have
migrations after release if the code breaks without the migration.

So I propose to put at least the migration already in the 2.9 code base
for the initial release...

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by bgeuken** [comment]
@adrianschroeter Given that we have not yet released 2.9 we can do that :confused: . Though I am a bit puzzled how you imagine this in general. Since having db migrations is a given in the rails world.
We would limit ourself quite a lot with such a rule, eg. how do we introduce the azure upload to 2.9 if we can not have database migrations?

**Comment by adrianschroeter** [comment]
On Dienstag, 13. März 2018, 10:54:58 CET wrote Björn Geuken:
> @adrianschroeter Given that we have not yet released 2.9 we can do that :confused:

yes, that was my point. Either do it now or we would need to adapt the code with fallbacks.

> . Though I am a bit puzzled how you imagine this in general. Since having db migrations is a given in the rails world.
> We would limit ourself quite a lot with such a rule, eg. how do we introduce the azure upload to 2.9 if we can not have database migrations?

We need to maintain a stable base we do this for more >10 years this way. rails is just an implementation
and can not dictate how admins maintain their servers. 

The new azure upload feature may not work without the migration, but it must not break anything
else, if the migration did not run. It is okay to require to run migrations for new functionality.

In any case, it must be save to update the packages from the stable base automated.
And given that running migrations is risky and not error free we can not do it autmatically.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by bgeuken** [comment]
@adrianschroeter I am fine with cherry picking the migration in this particular case. But I don't get this constrain that you set now.
And I can't follow why doing database migrations would result in an unstable base.

But this seems to be a more general discussion ad we should discuss this on our mailing list. So that everyone can get involved.

File: src/api/app/models/configuration.rb [hunk 1]
```diff
@@ -82,13 +82,16 @@ def amqp_namespace
...
+  def passwords_changable?(user = nil)
+    change_password && CONFIG['proxy_auth_mode'] != :on && (user.try(:ignore_auth_services?) || CONFIG['ldap_mode'] != :on)
   end
 
-  def accounts_editable?
-    (CONFIG['proxy_auth_mode'] != :on || (CONFIG['proxy_auth_mode'] == :on && CONFIG['proxy_auth_account_page'].present?)) && \
-      CONFIG['ldap_mode'] != :on
+  def accounts_editable?(user = nil)
+    (
+      CONFIG['proxy_auth_mode'] != :on || CONFIG['proxy_auth_account_page'].present?
```
Comment (ChrisBr) [inline src/api/app/models/configuration.rb]: Why did you change
```
(CONFIG['proxy_auth_mode'] == :on && CONFIG['proxy_auth_account_page'].present?)
```

File: src/api/app/views/models/_user.xml.builder [hunk 2]
```diff
@@ -13,6 +13,8 @@ xml.person do
     xml.globalrole(role.title)
   end
 
+  xml.ignore_auth_services(my_model.ignore_auth_services)
```
Comment (ChrisBr) [inline src/api/app/views/models/_user.xml.builder]: Is it necessary that we expose that in the XML?

File: src/api/app/views/webui/user/_dropdown_menu.html.haml [hunk 3]
```diff
@@ -20,7 +20,12 @@
     %li
       = sprite_tag('cog_add')
       = link_to('Make user admin', user_update_path(user: { login: user.login, mark_as_admin: true }), |
-              method: :patch, data: { confirm: "Are you sure?" }) |
+                method: :patch, data: { confirm: "Are you sure?" }) |
+    - if ::Configuration.passwords_changable?(user)
```
Comment (ChrisBr) [inline src/api/app/views/webui/user/_dropdown_menu.html.haml]: That would mean that you can also set a user to local user in Proxy Mode? That is not stated in the card. Do we want that?

File: src/api/spec/controllers/person_controller_spec.rb [hunk 4]
```diff
@@ -44,13 +61,29 @@
...
+        it { expect(old_password).to eq(user.password_digest) }
+      end
+
+      context 'and the user is configured to authorize locally' do
+        before do
+          user.update(ignore_auth_services: true)
+          post :post_userinfo, params: { login: user.login, cmd: 'change_password', format: :xml }
+        end
+
+        it { expect(old_password).to_not eq(user.reload.password_digest) }
```
Comment (ChrisBr) [inline src/api/spec/controllers/person_controller_spec.rb]: This test would always pass as the password_digest is the has and would never match the old_password. Not sure if it makes sense to compare the old password digest with the new?

File: src/api/spec/controllers/webui/user_controller_spec.rb [hunk 5]
```diff
@@ -185,7 +185,12 @@
         login admin_user
         post :save, params: {
           user: {
-            login: user.login, realname: 'another real name', email: 'new_valid@email.es', state: 'locked', role_ids: new_global_roles.pluck(:id)
+            login:                user.login,
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/user_controller_spec.rb]: Can we also test that a normal user can not update the ``ignore_auth_services`` field?

File: src/api/app/models/configuration.rb (hunk 1 above)
Comment (bgeuken) [reply src/api/app/models/configuration.rb]: @ChrisBr Because the second check (`|| CONFIG['proxy_auth_mode'] == :on `) was redundant. If 'proxy_auth_mode' is not :on, we can assume it's :on.

File: src/api/app/views/models/_user.xml.builder (hunk 2 above)
Comment (bgeuken) [reply src/api/app/views/models/_user.xml.builder]: Why should we hide it? I wouldn't see this as data we need to hide. Though we could only show it in LDAP mode though. Or only for admins.

File: src/api/app/views/webui/user/_dropdown_menu.html.haml (hunk 3 above)
Comment (bgeuken) [reply src/api/app/views/webui/user/_dropdown_menu.html.haml]: IIRC, we talked about a generic solution in the planning. But I can change it.

File: src/api/spec/controllers/person_controller_spec.rb (hunk 4 above)
Comment (bgeuken) [reply src/api/spec/controllers/person_controller_spec.rb]: Not sure what you mean, tbh. The digest is a hash based on the clear text password. So if the password stays the same the hash is the same as well.  Note that there are also tests for the case that the password does not change.

File: src/api/spec/controllers/webui/user_controller_spec.rb (hunk 5 above)
Comment (bgeuken) [reply src/api/spec/controllers/webui/user_controller_spec.rb]: Good idea :+1: Added it now.

File: src/api/app/views/models/_user.xml.builder (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/views/models/_user.xml.builder]: Not that we need to hide it but for the majority of users this information will be useless. I would limit it to admins!

File: src/api/spec/controllers/person_controller_spec.rb (hunk 4 above)
Comment (ChrisBr) [reply src/api/spec/controllers/person_controller_spec.rb]: Then please call it ``old_password_digest`` because if you read it, it looks like you compare the password with the hash.
